#### Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you might choose one over the other.

Both techniques are used for converting  categorical variable into numerical values.
Ordinal encoding is used based on the rank whereas Label encoding used to assign numerical value to categorical variable.

example:

if we are working with a dataset that contains a variable for "education level" with categories "high school", "college", "graduate school", and "PhD", we might choose to use ordinal encoding to capture the hierarchy between these categories. However, if we are working with a dataset that contains a variable for "car make" with categories "Toyota", "Honda", "Ford", and "Chevrolet", we might choose to use label encoding, as there is no clear order or hierarchy between these categories.

#### Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in a machine learning project.

Target guided ordinal encoding used to assign numerical value to categorical variable based on the target variable. and assigning is done base on mean and median.

example:

Suppose we have a dataset that contains a categorical variable "gender" with values "male" and "female", and a binary target variable "loan_status" with values 0 and 1, where 0 indicates a loan rejection and 1 indicates a loan approval. We want to encode the "gender" variable using Target Guided Ordinal Encoding

#### Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

Covariance represents extent that show how two variable are dependent. It is very important concept in statistical analysis as it is helps to understand the relationships between two variable.

cov(X, Y) = (1/n) ∑(Xi - X̄)(Yi - Ȳ)

#### Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium, large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library. Show your code and explain the output.

In [ ]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, OrdinalEncoder
import warnings
warnings.simplefilter('ignore')

#### 1.OneHotEncoding

In [ ]:
df = pd.DataFrame({
    'colors' : ['red','green','blue']
})

In [ ]:
df

colors
0    red
1  green
2   blue

In [ ]:
oneencode = OneHotEncoder()

In [ ]:
encoded = oneencode.fit_transform(df[['colors']]).toarray()

In [ ]:
encoded_df = pd.DataFrame(encoded,columns=oneencode.get_feature_names_out())

In [ ]:
encoded_df

colors_blue  colors_green  colors_red
0          0.0           0.0         1.0
1          0.0           1.0         0.0
2          1.0           0.0         0.0

#### 2. Ordinal Encoding

In [ ]:
df = pd.DataFrame({
    'size' : ['small','medium','large']
})

In [ ]:
df

size
0   small
1  medium
2   large

In [ ]:
ord_encoder = OrdinalEncoder(categories=[['small','medium','large']])

In [ ]:
encoded = ord_encoder.fit_transform(df[['size']])

In [ ]:
encoded

array([[0.],
       [1.],
       [2.]])

#### 3. Label Encoding

In [ ]:
df = pd.DataFrame({
    'material' : ['wood','metal','plastic']
})

In [ ]:
lbl_encoder = LabelEncoder()

In [ ]:
encoded = lbl_encoder.fit_transform(df[['material']])

In [ ]:
encoded

array([2, 0, 1])

#### Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education level. Interpret the results.

In [ ]:
df = pd.DataFrame({
    'Age' : [21,23,25,30],
    'Income' : [12000,50000,45000,67000],
    'Education level': ['Bcom','Btech','Arts','Bscs']
})

In [ ]:
df.cov()

Age        Income
Age        14.916667  7.683333e+04
Income  76833.333333  5.296667e+08

#### Q6. You are working on a machine learning project with a dataset containing several categorical variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD), and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for each variable, and why?

As you can see all are categorical variable for gender and employment status i would use one hot encoding and for education level i would use ordinal encoding.

#### Gender

In [ ]:
df = pd.DataFrame({
    'Gender' : ['Male','Female']
})

In [ ]:
df

Gender
0    Male
1  Female

In [ ]:
one_encoder = OneHotEncoder()

In [ ]:
encoded = one_encoder.fit_transform(df[['Gender']]).toarray()

In [ ]:
encoded_df = pd.DataFrame(encoded,columns=one_encoder.get_feature_names_out())

In [ ]:
encoded_df

Gender_Female  Gender_Male
0            0.0          1.0
1            1.0          0.0

#### Education level

In [ ]:
df = pd.DataFrame({
    'Education level': ['High school','Bachelor','Master','Phd']
})

In [ ]:
df

Education level
0     High school
1        Bachelor
2          Master
3             Phd

In [ ]:
ord_encoder = OrdinalEncoder(categories=[['High school','Bachelor','Master','Phd']])

In [ ]:
encoded = ord_encoder.fit_transform(df[['Education level']])

In [ ]:
encoded

array([[0.],
       [1.],
       [2.],
       [3.]])

#### Employment status

In [ ]:
df = pd.DataFrame({
    'Employment status' : ['part time','full time']
})

In [ ]:
df

Employment status
0         part time
1         full time

In [ ]:
one_encoder = OneHotEncoder()

In [ ]:
encoded = one_encoder.fit_transform(df[['Employment status']]).toarray()

In [ ]:
encoded_df = pd.DataFrame(encoded,columns=one_encoder.get_feature_names_out())

In [ ]:
encoded_df

Employment status_full time  Employment status_part time
0                          0.0                          1.0
1                          1.0                          0.0

#### Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/East/West). Calculate the covariance between each pair of variables and interpret the results.

In [ ]:
import numpy as np
import pandas as pd

# Set seed for reproducibility
np.random.seed(321)

# Generate data
n = 1000
temp = np.random.normal(25, 5, n)
humidity = np.random.normal(60, 10, n)
weather_condition = np.random.choice(['Sunny', 'Cloudy', 'Rainy'], size=n)
wind_direction = np.random.choice(['North', 'South', 'East', 'West'], size=n)

# Create dataframe
df = pd.DataFrame({
    'Temperature': temp,
    'Humidity': humidity,
    'Weather Condition': weather_condition,
    'Wind Direction': wind_direction
})

# Show first few rows
df.head()

Temperature   Humidity Weather Condition Wind Direction
0    25.862597  50.526311             Sunny          South
1    33.177413  55.809608             Sunny          South
2    25.186682  70.091030             Sunny           West
3    20.579252  68.981094             Sunny          South
4    19.284039  78.624127             Rainy           East

#### Calculating Covariance Matrix for Numerical Variables only

In [ ]:
df.cov(numeric_only=True)

Temperature    Humidity
Temperature    25.165416    1.610779
Humidity        1.610779  105.612893

The covariance between "Temperature" and "Humidity" is 1.611 , indicating a positive relationship between the two variables. This means that as temperature increases, humidity tends to increase as well. The variances of each variable are shown on the diagonal, with Humidity having a larger variance than Temperature.

To calculate the covariance between the continuous variables and the categorical variables, we can group the data by the categorical variables and calculate the covariance for each group. Here's an example code:

It is important to note that we cannot calculate the covariance between continuous and categorical variables since covariance requires numerical data. Therefore, we cannot interpret the covariance between "Temperature" and "Weather Condition" or between "Humidity" and "Wind Direction". In general, we need to be careful when interpreting covariance and consider the nature of the variables being analyzed.

ANOVA Should be used to compare significance of Categorical variables with Numeric Variables